design/20210203.certificate-request-identity.md

```diff
@@ -0,0 +1,165 @@
+resources.
+
+The webhook will be responsible for enforcing the following during a CREATE
+operation:
+- No identity fields have been set by the user creating the resource
+- Set identity fields to exactly what is received from the API server in the
+    [UserInfo](https://github.com/kubernetes/api/blob/master/authentication/v1/types.go#L102)
+
+The webhook will also responsible for enforcing the following during an UPDATE
+operation:
+- No changes to identity fields are allowed to be made
+
+### API Changes
+
+In order to expose the identity of who created `CertificateRequest` resources,
+these resources must be updated to have parity with the upstream Kubernetes
+certificates API.
+This means that the `CertificateRequest` API type be updated to include the
+following fields in spec, for all API versions:
+
+```go
+	// Username contains the name of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	Username string `json:"username,omitempty"`
+	// UID contains the uid of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	UID string `json:"uid,omitempty"`
+	// Groups contains group membership of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +listType=atomic
+	// +optional
+	Groups []string `json:"groups,omitempty"`
+	// Extra contains extra attributes of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	Extra map[string][]string `json:"extra,omitempty"`
+```
+
+### Upgrading
+
+Any `CertificateRequest`s that are present in the cluster at the time of
+upgrade, won't and will never have their identity fields populated. This should
+be acceptable.
+
+All subsequent `CertificateRequest`s created after this upgrade will have their
+identities populated.
+
+### Test Plan
+
+Unit tests will be created is ensure that the [properties](#proposal) are
+enforced.
+
+End to end tests will ensure that different users creating `CertificateRequest`
+resources have their respective identities populated in the fields.
+
+### Risks and Mitigations
+
+There are large security implications if the identity fields that other
+components rely on are wrong, or could be made fraudulent. Special care when
+testing needs to be given to ensure the properties described above are correct
+and enforced.
+
+### Certificate Identity Pass Down
+
+`CertificateRequest`s created and managed via `Certificate` resources will have
+the identity of the cert-manager controller. Though this design doesn't provide
+a solution to components that need to traverse the identity to the original
+user who created the `Certificate` resource, some possible further design
+options are;
+- cert-manager would override the identity fields with the of the identity that
```

Comment:
If cert-manager was to update these fields after the CertificateRequest has been created, would the webhook not reject the update (as it enforces immutability of identity fields)?

Reply:
The fields that are added happen at admission time, effectively the same CREATE/UPDATE operation when sent to the API server. The mutating webhook doesn't make an update during its evaluation itself, but instead will respond to the API server with a JSON patch which is applied during the object creation/update internally to the API server.